### ISSUE TITLE
NullableToStringCall reports on "${this.javaClass.simpleName} ..."

### ISSUE THREAD

**Issue by inktomi** [detekt/detekt#4059]
This is related to the discussion in https://github.com/detekt/detekt/issues/3855 but isn't quite what the main ticket is reporting so I wanted to call this out. We can combine or close this if it's not appropriate here as a standalone issue. 

## Expected Behavior of the rule

`simpleName`, as defined in Java 11's `Class`, should never return null. The code doens't appear to have a path which would allow it. Similarly, `javaClass` defined in Kotlin's `JvmClassMapping` also cannot return null, and so the combined call to `javaClass.simpleName` should never result in a null value. 

Despite this, the `NullableToStringCall` check is reporting that there's a nullability issue in this string. 

## Context

The `NullableToStringCall` rule should not report on Java values which cannot be null.

**Comment by cortinico** [comment]
Thanks for the report and for providing context.

> The code doens't appear to have a path which would allow it.

Detekt is not analyzing the function code, therefore can't know that there are no paths to return `null`. I'm unsure if this API is annotate with `@Nullable` or `@NonNull` or similar. If not, detekt will consider this as an API that could return `null` and will mark that issue.

I believe the best would be to provide the exclusion mechanism requested in #3855

**Comment by BraisGabin** [comment]
I can confirm this issue. Here a test to reproduce it: 
```kotlin
describe("Issue #4059") {
    it("We shouldn't fail") {
        val dolar = "$"
        val code = """
            class A {
                fun b() {
                    println("$dolar{javaClass.simpleName}")
                }
            }
        """
        val actual = NullableToStringCall().compileAndLintWithContext(env, code)
        assertThat(actual).isEmpty()
    }
}
```

And as far as I can see IntelliJ detects that this function never returns null:
![Captura de pantalla 2021-10-02 a las 14 26 08](https://user-images.githubusercontent.com/721244/135716013-e79a9a2a-e736-4a04-8239-27d1b515a195.png)

And it marks as a warnign to use `!!` on it because it's already non-null:
![Captura de pantalla 2021-10-02 a las 14 27 25](https://user-images.githubusercontent.com/721244/135716044-008355ac-a0d9-4557-aecd-d0c5f43b7831.png)

But this issue is not related with this rule. I can see the same issue in `UnnecessaryNotNullOperator`. And here a test to prove it:

```kotlin
describe("Issue #4059") {
    it("We shouldn't fail") {
        val code = """
            class A {
                fun b() {
                    val a = javaClass.simpleName!!
                }
            }
        """
        val actual = UnnecessaryNotNullOperator().compileAndLintWithContext(env, code)
        assertThat(actual).hasSize(1)
    }
}
```

So it seems that our `BindingContext` isn't getting the proper data.

**Comment by 3flex** [comment]
I've looked into this.

Essentially the problem is that IntelliJ supports external annotations: https://www.jetbrains.com/help/idea/external-annotations.html

But the Kotlin compiler does not. It used to, but support was removed. So detekt can't get nullability info from the compiler for things like the JDK which aren't annotated directly.

I'm not sure what the best path forward is. We could add support for external annotations to detekt itself, but it seems clunky. We could also assume for rules like this one that platform types are not nullable which will avoid false positives.

**Comment by BraisGabin** [comment]
Nice finding. How hard would it be to add support for external annotations? Could it be done in the core? I mean, adding that information inside the `ContextBinding` so the rules doesn't need to care about this. Depending on those I would suggest this:

- If it is kind of easy and it can be done in the core: Go for it.
- else:  Create a new issue to talk abot it. And, to fix this issue, ignore platform types in this rule.
